File: src/Web/KRT.Web/src/app/shared/utils/brcode-parser.ts
/**
 * Parser de BRCode (PIX Copia e Cola) no formato EMV QR Code.
 *
 * Formato TLV: cada campo = 2 chars (tag) + 2 chars (tamanho) + N chars (valor)
 * Tags compostas (26, 62) contêm sub-TLVs no mesmo formato.
 */

export interface BRCodeData {
  pixKey: string;
  amount: number;
  merchantName: string;
  city: string;
  txId: string;
  isValid: boolean;
}

/** Parse TLV (Tag-Length-Value) fields from an EMV string */
function parseTLV(data: string): Map<string, string> {
  const result = new Map<string, string>();
  let i = 0;
  while (i + 4 <= data.length) {
    const tag = data.substring(i, i + 2);
    const lenStr = data.substring(i + 2, i + 4);
    const len = parseInt(lenStr, 10);
    if (isNaN(len) || len < 0 || i + 4 + len > data.length) break;
    result.set(tag, data.substring(i + 4, i + 4 + len));
    i += 4 + len;
  }
  return result;
}

/** Parse a PIX BRCode (EMV payload) and extract payment data */
export function parseBRCode(payload: string): BRCodeData {
  const invalid: BRCodeData = {
    pixKey: '',
    amount: 0,
    merchantName: '',
    city: '',
    txId: '',
    isValid: false,
  };

  if (!payload || payload.trim().length < 20) return invalid;

  try {
    const clean = payload.trim();
    const tags = parseTLV(clean);

    // Tag 00: Payload Format Indicator (deve ser "01")
    if (tags.get('00') !== '01') return invalid;

    // Tag 26: Merchant Account Information (contém subtags)
    const tag26 = tags.get('26');
    if (!tag26) return invalid;
    const sub26 = parseTLV(tag26);

    // Subtag 00: GUI deve ser "BR.GOV.BCB.PIX"
    if (sub26.get('00') !== 'BR.GOV.BCB.PIX') return invalid;

    // Subtag 01: Chave PIX
    const pixKey = sub26.get('01') || '';

    // Tag 54: Valor da transação
    const amountStr = tags.get('54') || '0';
    const amount = parseFloat(amountStr) || 0;

    // Tag 59: Nome do recebedor
    const merchantName = tags.get('59') || '';

    // Tag 60: Cidade
    const city = tags.get('60') || '';

    // Tag 62: Dados adicionais (contém subtags)
    let txId = '';
    const tag62 = tags.get('62');
    if (tag62) {
      const sub62 = parseTLV(tag62);
      txId = sub62.get('05') || '';
    }

    return {
      pixKey,
      amount,
      merchantName,
      city,
      txId,
      isValid: pixKey.length > 0 && amount > 0,
    };
  } catch {
    return invalid;
  }
}

File: src/Services/KRT.Payments/KRT.Payments.Api/Services/BRCodeParser.cs
using System.Globalization;

namespace KRT.Payments.Api.Services;

public class BRCodeData
{
    public decimal Amount { get; set; }
    public string PixKey { get; set; } = "";
    public string MerchantName { get; set; } = "";
    public string MerchantCity { get; set; } = "";
    public string TxId { get; set; } = "";
    public bool IsValid { get; set; }
}

/// <summary>
/// Parser de BRCode (PIX Copia e Cola) no formato EMV QR Code.
/// Formato TLV: cada campo = 2 chars (tag) + 2 chars (tamanho) + N chars (valor).
/// Tags compostas (26, 62) contêm sub-TLVs no mesmo formato.
/// </summary>
public class BRCodeParser
{
    public BRCodeData Parse(string brcode)
    {
        var result = new BRCodeData();

        if (string.IsNullOrWhiteSpace(brcode) || brcode.Trim().Length < 20)
            return result;

        try
        {
            var clean = brcode.Trim();
            var tags = ParseTLV(clean);

            // Tag 00: Payload Format Indicator (deve ser "01")
            if (!tags.TryGetValue("00", out var pfi) || pfi != "01")
                return result;

            // Tag 26: Merchant Account Information (contém subtags)
            if (!tags.TryGetValue("26", out var tag26))
                return result;

            var sub26 = ParseTLV(tag26);

            // Subtag 00: GUI deve ser "BR.GOV.BCB.PIX"
            if (!sub26.TryGetValue("00", out var gui) || gui != "BR.GOV.BCB.PIX")
                return result;

            // Subtag 01: Chave PIX
            if (sub26.TryGetValue("01", out var pixKey))
                result.PixKey = pixKey;

            // Tag 54: Valor da transação
            if (tags.TryGetValue("54", out var amountStr) &&
                decimal.TryParse(amountStr, NumberStyles.Any, CultureInfo.InvariantCulture, out var amount))
                result.Amount = amount;

            // Tag 59: Nome do recebedor
            if (tags.TryGetValue("59", out var merchantName))
                result.MerchantName = merchantName;

            // Tag 60: Cidade
            if (tags.TryGetValue("60", out var city))
                result.MerchantCity = city;

            // Tag 62: Dados adicionais (contém subtags)
            if (tags.TryGetValue("62", out var tag62))
            {
                var sub62 = ParseTLV(tag62);
                if (sub62.TryGetValue("05", out var txId))
                    result.TxId = txId;
            }

            result.IsValid = !string.IsNullOrEmpty(result.PixKey) && result.Amount > 0;
        }
        catch
        {
            // Parsing falhou — retorna resultado vazio (IsValid = false)
        }

        return result;
    }

    private static Dictionary<string, string> ParseTLV(string data)
    {
        var result = new Dictionary<string, string>();
        var i = 0;

        while (i + 4 <= data.Length)
        {
            var tag = data.Substring(i, 2);
            var lenStr = data.Substring(i + 2, 2);

            if (!int.TryParse(lenStr, out var len) || len < 0 || i + 4 + len > data.Length)
                break;

            result[tag] = data.Substring(i + 4, len);
            i += 4 + len;
        }

        return result;
    }
}
